morph_tool/morphdb.py

```diff
@@ -0,0 +1,257 @@
+'''Provides two classes to work with neurondb files
+
+The main use is to provide a neuronDB loader and a method to retrieve
+information as a dataframe:
+
+MorphologyDB('neurondb.xml').df
+'''
+import json
+import logging
+from pathlib import Path
+from typing import Any, Callable, Dict, Iterable, List, Optional
+
+import pandas as pd
+import xmltodict
+
+L = logging.getLogger(__name__)
+
+NEURONDB_XML = 'neuronDB.xml'
+MTYPE_MSUBTYPE_SEPARATOR = ':'
+
+
+class MorphInfo:
+    '''A class the contains information about a morphology.
+
+    Its role is to abstract away the raw data.
+    '''
+
+    # Attributes found in the <repair></repair> block of the XML
+    BOOLEAN_REPAIR_ATTRS = [
+        'use_axon',
+        'use_dendrites',
+        'axon_repair',
+        'dendrite_repair',
+        'basal_dendrite_repair',
+        'tuft_dendrite_repair',
+        'oblique_dendrite_repair',
+        'unravel',
+        'use_for_stats',
+    ]
+
+    COLUMNS = ['name', 'mtype', 'msubtype', 'fullmtype',
+               'layer'] + BOOLEAN_REPAIR_ATTRS + ['axon_inputs']
+
+    def __init__(self, item: Dict[str, Any]):
+        '''MorphInfo ctor.
+
+        Args:
+            item: A dictionnary that represents the content of the XML file.
+                  The only mandatory keys are [name, mtype, layer]
+        '''
+        self.item = item
+        self.name = item['name']
+        self.mtype = item['mtype']
+        self.layer = str(item['layer'])
+        self.msubtype = item.get('msubtype') or ''
+        self.fullmtype = MTYPE_MSUBTYPE_SEPARATOR.join(filter(None, [self.mtype, self.msubtype]))
+
+        def is_true(el: Optional[str]) -> bool:
+            '''Parse a string representing a boolean repair flag and returns its boolean value
+
+            Unless clearly stated as false, missing tags default to True
+
+            - According to Eilif, an empty use_axon (corresponding to a null in the database)
+              means that the axon is supposed to be used
+
+            - dendrite_repair       defaults to True in BlueRepairSDK
+            - basal_dendrite_repair defaults to True in BlueRepairSDK
+            - unravel: well I guess we always want to do it
+            '''
+            return el in (None, '', 'true', 'True', )
+
+        repair = item.get('repair', {})
+        for attr in MorphInfo.BOOLEAN_REPAIR_ATTRS:
+            setattr(self, attr, is_true(repair.get(attr)))
+
+        # Case where there is a single <axoninput></axoninput> tag in the XML
+        self.axon_inputs = repair.get('axon_sources', {}).get('axoninput', [])
+        if not isinstance(self.axon_inputs, list):
+            self.axon_inputs = [self.axon_inputs]
+
+        # lineage information
+        self.dendrite_donor = item.get('parent') or item.get('dendrite')
+        self.axon_donor = item.get('parent') or item.get('axon')
+
+    def __hash__(self):
```

Comment:
why do we consider the hash to be the same when the subtypes/donors/etc differ?

Reply:
This is based on `add_morphdb` and this is how it is decided if a morphology should be added to the morphdb or not.
```python
    def add_morph(self, name, mtype, layer, parent=None, dendrite=None, axon=None):
        '''add a morphology based on name, mtype and layer

        Add lineage information: if parent, then dendrite and axon are both set,
        or they are set independently

        If the name, mtype, layer combination already exists, then nothing is added

        Return:
            bool: True if morphology added, False otherwise
        '''
        morph_info = XMLMorphInfo(name=name, mtype=mtype, layer=layer,
                                  use_axon=None, use_dendrites=None,
                                  raw_xml=None)
        morph_key = MorphologyDB._morph_key(morph_info)
        if morph_key in self._known_morphologies:
            return False
```
However, after double checking, the mtype used is in fact the fullmtype so I will update the PR with that.

Reply:
Maybe that's a bad design decision though. The logic involved in `MorphologyDB.add_morph` should not leak into designing the `MorphInfo` class.